spring-kafka/src/main/java/org/springframework/kafka/core/reactive/ReactiveKafkaProducerTemplate.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.core.reactive;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.reactivestreams.Publisher;
+
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.kafka.support.KafkaHeaders;
+import org.springframework.kafka.support.converter.MessagingMessageConverter;
+import org.springframework.kafka.support.converter.RecordMessageConverter;
+import org.springframework.messaging.Message;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.kafka.sender.KafkaSender;
+import reactor.kafka.sender.SenderOptions;
+import reactor.kafka.sender.SenderRecord;
+import reactor.kafka.sender.SenderResult;
+import reactor.kafka.sender.TransactionManager;
+
+/**
+ * Reactive kafka producer operations implementation.
+ *
+ * @param <K> the key type.
+ * @param <V> the value type.
+ *
+ * @author Mark Norkin
+ */
+public class ReactiveKafkaProducerTemplate<K, V> implements ReactiveKafkaProducerOperations<K, V>, AutoCloseable, DisposableBean {
+	private final KafkaSender<K, V> sender;
+	private final RecordMessageConverter messageConverter;
+
+	public ReactiveKafkaProducerTemplate(SenderOptions<K, V> senderOptions) {
+		this(senderOptions, new MessagingMessageConverter());
+	}
+
+	public ReactiveKafkaProducerTemplate(SenderOptions<K, V> senderOptions, RecordMessageConverter messageConverter) {
+		this.sender = KafkaSender.create(senderOptions);
+		this.messageConverter = messageConverter;
+	}
+
+	@Override
+	public <T> Flux<SenderResult<T>> sendTransactionally(Publisher<? extends SenderRecord<K, V, T>> records) {
+		Flux<Flux<SenderResult<T>>> sendTransactionally = this.sender.sendTransactionally(Flux.just(records));
```

Comment:
Don't you mean here `Flux.from(records)`?

Reply:
Well, Flux.from can't be used here as  `sendTransactionally` accepts publisher of publishers e.g. `Publisher<? extends Publisher<? extends SenderRecord<K, V, T>>> records`, but Flux.from would return `Flux<? extends SenderRecord<K, V, T>>` instead of `Flux<? extends Publisher<? extends SenderRecord<K, V, T>>>`